dbt/dbt-run.py

```diff
@@ -133,10 +150,94 @@ def run_incremental_models():
+      )
+      conn.commit()
+
+def get_last_processed_timestamp():
+  with connection() as conn:
+    with conn.cursor() as cur:
+      cur.execute(f"""
+          SELECT MAX(timestamp)
+          FROM {BATCH_STATUS_TABLE}
+          WHERE status = 'success'
+      """)
+      result = cur.fetchone()
+      if result and result[0]:
+        return result[0]
+      return '1970-01-01 00:00:00'
+    
+def get_max_timestamp():
+  with connection() as conn:
+    with conn.cursor() as cur:
+      cur.execute(f"""
+          SELECT MAX(saved_timestamp)
+          FROM {SCHEMA_NAME}.{METADATA_TABLE_NAME}
+      """)
+      return cur.fetchone()[0]
+    
+def handle_batch_update_error(last_processed_timestamp, dataemon_interval):
+   print(f"Error running dbt for batch {last_processed_timestamp}")
+   update_batch_status(last_processed_timestamp, "error")
+   time.sleep(dataemon_interval)
+
+def handle_batch_update_success(last_processed_timestamp, dataemon_interval, max_timestamp):
+  update_batch_status(last_processed_timestamp, "success")
+
+  if max_timestamp == last_processed_timestamp:
+     time.sleep(dataemon_interval)
+
+def check_table_exists(schema_name, table_name):
+    with connection() as conn:
+        with conn.cursor() as cur:
+            cur.execute(f"""
+                SELECT EXISTS (
+                    SELECT FROM information_schema.tables 
+                    WHERE table_schema = '{schema_name}' AND table_name = '{table_name}'
+                )
+            """)
+            return cur.fetchone()[0]
+   
+def run_dbt_in_batches():
+  last_processed_timestamp = get_last_processed_timestamp()
+  batch_size = int(os.getenv("DBT_BATCH_SIZE") or 10000)
+  dataemon_interval = int(os.getenv("DATAEMON_INTERVAL") or 5)
+
+  while True:
+     update_dbt_deps()
+     result = subprocess.run([
+        "dbt", "run",
+        "--profiles-dir", ".dbt",
+        "--vars", f'{{start_timestamp: "{last_processed_timestamp}", batch_size: {batch_size}}}',
+        "--exclude", "config.materialized:view"
+      ])
+     
+     if result.returncode != 0:
+        handle_batch_update_error(last_processed_timestamp, dataemon_interval)
+        continue
+     
+     max_timestamp = get_max_timestamp()
+     handle_batch_update_success(last_processed_timestamp, dataemon_interval, max_timestamp)
+     
+     last_processed_timestamp = max_timestamp
+
+def run_dbt():
+  while True:
+      update_models()
+      run_incremental_models()
+      time.sleep(int(os.getenv("DATAEMON_INTERVAL") or 5))
 
 if __name__ == "__main__":
+  print("Starting dbt run")
   setup()
-  while True:
-    update_models()
-    run_incremental_models()
-    time.sleep(int(os.getenv("DATAEMON_INTERVAL") or 5))
+  if os.getenv("RUN_DBT_IN_BATCHES"):
+    if not check_table_exists(SCHEMA_NAME, METADATA_TABLE_NAME):
+       raise psycopg2.errors.UndefinedTable(f"The table {METADATA_TABLE_NAME} does not exist in the database.")
+    run_dbt_in_batches()
+  else:
+    run_dbt()
```

Comment:
This is great! I have another question though: 

what happens if a cht-sync instance is started with `RUN_DBT_IN_BATCHES=false`, and then, later, when there's a large influx of docs, the cht-sync is restarted with `RUN_DBT_IN_BATCHES=true`. Will this make DBT sync process everything again because we don't have a batch status stored?

Reply:
I have removed the `document_metadata` table check because there was a bug where dbt would not run for new deployments where the tables ans views are not yet created. The dbt ls [command](https://docs.getdbt.com/reference/commands/list) should be able to help identify if the model is defined but I didn't manage to get it to work so we'll catch that error elsewhere in the code.

> what happens if a cht-sync instance is started with RUN_DBT_IN_BATCHES=false, and then, later, when there's a large influx of docs, the cht-sync is restarted with RUN_DBT_IN_BATCHES=true. Will this make DBT sync process everything again because we don't have a batch status stored?

I have added a check on the document_metadata table for the latest timestamp that handles this scenario.